### PR TITLE
feat(backend): add BTC balance call

### PR DESCRIPTION
**Pull request by AntonioVentilii-DFINITY**
# Motivation

We add the backend call to fetch the BTC balance of a generic address, based on the [IC documentation](https://internetcomputer.org/docs/current/developer-docs/multi-chain/bitcoin/using-btc/read-state#reading-current-balance).

### CODE REVIEW

File: src/backend/src/lib.rs [hunk 1]
```diff
@@ -265,6 +266,12 @@ async fn caller_btc_address(network: BitcoinNetwork) -> String {
     public_key_to_p2pkh_address(network, &ecdsa_pubkey_of(&ic_cdk::caller()).await)
 }
 
+/// Returns the balance of the given Bitcoin address.
+#[query]
```
Comment (peterpeterparker) [inline src/backend/src/lib.rs:270]: Public is fine?

File: src/backend/src/lib.rs [hunk 2]
```diff
@@ -265,6 +266,12 @@ async fn caller_btc_address(network: BitcoinNetwork) -> String {
...
+async fn btc_balance(address: String, network: BitcoinNetwork) -> u64 {
+    bitcoin_lib::get_balance(network, address).await
```
Comment (peterpeterparker) [inline src/backend/src/lib.rs:272]: nth: just `get_balance` and declare the `use` at the top for consistency?

File: src/backend/src/bitcoin_lib.rs [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+/// See [IC Bitcoin Documentation](https://internetcomputer.org/docs/current/references/ic-interface-spec/#ic-bitcoin_get_balance)
+pub async fn get_balance(network: BitcoinNetwork, address: String) -> u64 {
+    let balance_res = bitcoin_get_balance(GetBalanceRequest {
+        address,
+        network,
+        min_confirmations: None,
+    })
+    .await;
+
+    balance_res.unwrap().0
```
Comment (peterpeterparker) [inline src/backend/src/bitcoin_lib.rs:17]: Shouldn't we properly catch the error or bubble the error to the caller?

File: src/backend/src/bitcoin_lib.rs [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+
+/// Returns the balance of the given bitcoin address.
+///
+/// Relies on the `bitcoin_get_balance` endpoint.
+/// See [IC Bitcoin Documentation](https://internetcomputer.org/docs/current/references/ic-interface-spec/#ic-bitcoin_get_balance)
+pub async fn get_balance(network: BitcoinNetwork, address: String) -> u64 {
+    let balance_res = bitcoin_get_balance(GetBalanceRequest {
+        address,
+        network,
+        min_confirmations: None,
```
Comment (peterpeterparker) [inline src/backend/src/bitcoin_lib.rs:13]: None is correct?

File: src/backend/src/lib.rs (hunk 1 above)
Comment (peterpeterparker) [inline src/backend/src/lib.rs:270]: Maybe a stupid question: Query is correct? `bitcoin_get_balance` isn't an inter-canister update call?

**Review by peterpeterparker — COMMENTED** [review]
OH! Another important question maybe: why do we need that call in the backend of Oisy? Couldn't we just query the balance from the frontend as we already do for utxos in `bitcoin.api.ts`?

File: src/backend/src/lib.rs (hunk 2 above)
Comment (AntonioVentilii-DFINITY) [reply src/backend/src/lib.rs:272]: yes, agreed!

**Review by peterpeterparker — CHANGES_REQUESTED** [review]
As discussed with @AntonioVentilii-DFINITY, we do not need this PR as the calls can be made from the frontend. Antonio has started implementing this in ic-js (https://github.com/dfinity/ic-js/pull/701). So, just blocking the PR so that Max do not take time to review it as well.

**Comment by AntonioVentilii-DFINITY** [comment]
closing (see @peterpeterparker comment above)

**Comment by peterpeterparker** [comment]
Turns out this was incorrect. Query can indeed be made from the frontend but, update calls requires a backend call.

**Comment by AntonioVentilii-DFINITY** [comment]
Re-opened with PR https://github.com/dfinity/oisy-wallet/pull/2184

**Pull request closed without merging**